### PR TITLE
275 renewals summary

### PR DESCRIPTION
**Pull request by banukutlu**
#275

### CODE REVIEW

File: app/jobs/application_job.rb [hunk 1]
```diff
@@ -13,10 +13,11 @@ def symphony_client
       @symphony_client || SymphonyClient.new
     end
 
-    def badge(message: message)
+    def badge(message: message, badge_class: badge_class)
+      badge_class ||= 'success'
```
Comment (cdmo) [inline app/jobs/application_job.rb:17]:
```suggestion
    def badge(message: message, badge_class: 'success')
```

File: app/jobs/renew_checkout_job.rb [hunk 2]
```diff
@@ -24,7 +24,7 @@ def perform(resource:, item_key:, session_token:)
           renewal_count: checkout.renewal_count,
           due_date: due_date,
           status: checkout.status_human,
-          badge: badge(message: 'Successfully renewed')
+          badge: badge(message: 'Successfully renewed', badge_class: 'success')
```
Comment (cdmo) [inline app/jobs/renew_checkout_job.rb:27]: if the default arg for `badge` is `'success'` you don't need to define it here... right?

```suggestion
          badge: badge(message: 'Successfully renewed')
```

File: app/jobs/application_job.rb (hunk 1 above)
Comment (cdmo) [reply app/jobs/application_job.rb:17]: you know im sure this was intentional, like you want that second keyword arg to be explicit in the method call? which if so, totally fine.

**Review by cdmo — COMMENTED** [review]
So if I'm reading this right this PR adds checkout renewal summary notifications as well as a failure badge for renewals that fail? Fail badges aren't in place for holds though, right? That's intentional just like summary notifications for hold update actions not being there is intentional. Just checking.

**Comment by cdmo** [comment]
The interaction of renewing and how it updates looks very nice to me.

File: app/javascript/submission_handling/renew_checkout.js [hunk 3]
```diff
@@ -18,6 +20,38 @@ const updateCheckout = function (data) {
     }
 };
 
+const renderProcessing = function () {
+    const infoHTML = `<div class="alert alert-info" role="alert">
+                        <span>Renewals are in progress...</span>
+                      </div>`;
+
+    if (document.querySelector('.renewals-summary')) {
+        document.querySelector('.renewals-summary').innerHTML = infoHTML;
```
Comment (cdmo) [inline app/javascript/submission_handling/renew_checkout.js:29]: so this condition is when there is a stale (complete or even active) summary notification up and it'll get replaced/over-written by this new renewal batch action. right?

File: app/javascript/submission_handling/shared.js [hunk 4]
```diff
@@ -54,8 +52,5 @@ export const scrollToTop = () => {
 };
 
 export const clearBadges = () => {
-    const badges = document.getElementsByClassName('badge-success');
-    while (badges.length > defaultIndex) {
-        badges[defaultIndex].remove();
-    }
+    document.querySelectorAll('.badge-success, .badge-danger').forEach((badge) => badge.parentNode.removeChild(badge));
```
Comment (cdmo) [inline app/javascript/submission_handling/shared.js:55]: this is a nice refactor.

**Review by cdmo — DISMISSED** [review]
things getting a little hairy IMO in terms of lots of notifications to the user and things happening, plus moar javascript, BUT, the PR does make sense to me. 👍

File: app/jobs/application_job.rb (hunk 1 above)
Comment (cdmo) [reply app/jobs/application_job.rb:17]: right right 👍

File: app/javascript/submission_handling/renew_checkout.js (hunk 3 above)
Comment (banukutlu) [reply app/javascript/submission_handling/renew_checkout.js:29]: yeah, correct.

Comment (cdmo) [reply app/javascript/submission_handling/renew_checkout.js:29]: ok cool. do think the renaming of the function will help set understanding for future code readers

File: app/jobs/application_job.rb (hunk 1 above)
Comment (cdmo) [reply app/jobs/application_job.rb:17]: oh this reply is sort of in the wrong thread kinda

Comment (banukutlu) [reply app/jobs/application_job.rb:17]: wow weird, how did that happen!

Comment (banukutlu) [reply app/jobs/application_job.rb:17]: > you know im sure this was intentional, like you want that second keyword arg to be explicit in the method call? which if so, totally fine.

I do not recall any intention of why I did it like that, possibly after changing a bunch that is what I left with and forgot to clean it up, your suggestions makes more sense

Comment (cdmo) [reply app/jobs/application_job.rb]: really more of a code style type of review comment, not a big deal either way

File: app/javascript/submission_handling/polling.js [hunk 5]
```diff
@@ -75,14 +75,19 @@ const deleteData = function (jobId) {
...
+export const renderData = (target, resultCallback, otherRule = null, summaryCallback = null) => {
     pollFetch(target, otherRule).then((result) => {
         if (checkError(result)) {
             reportError(result.display_error);
         }
 
         resultCallback(result);
         deleteData(target);
+
+        if (summaryCallback && typeof summaryCallback === 'function') {
```
[the remark this answers is not in the document]
Comment (banukutlu) [reply app/javascript/submission_handling/polling.js:88]: added type checking to line 24 in the last commit

File: spec/features/checkouts_spec.rb [hunk 6]
```diff
@@ -22,24 +22,29 @@
...
 
-  context 'when patron renews a checkout successfully' do
+  context 'when patron renews a checkout successfully', js: true do
     before do
       page.check 'renewal_list__2145643:5:1'
       page.click_button 'Renew', match: :first
     end
 
-    it 'adds a success badge', js: true do
+    it 'adds a success badge' do
```
[the remark this answers is not in the document]
Comment (banukutlu) [reply spec/features/checkouts_spec.rb:31]: it does need it, moved it to context on line 25 :)

[the remark this answers is not in the document]
Comment (cdmo) [reply spec/features/checkouts_spec.rb:31]: oh ok, i see fair enough 👍

**Review by cdmo — DISMISSED** [review]

**Comment by cdmo** [comment]
Looks good... failing CI. mimemagic problem?

**Comment by banukutlu** [comment]
> Looks good... failing CI. mimemagic problem?

yes, it is #https://github.com/rails/rails/issues/41750, I am working on updating rails version. I guess I would do that in another branch, PR 🤔

**Comment by banukutlu** [comment]
> things getting a little hairy IMO in terms of lots of notifications to the user and things happening, plus moar javascript, BUT, the PR does make sense to me. 👍

totally agree, it is kinda getting too much moving parts for the user, and yes unfortunately moar JS :/

**Comment by banukutlu** [comment]
> So if I'm reading this right this PR adds checkout renewal summary notifications as well as a failure badge for renewals that fail? Fail badges aren't in place for holds though, right? That's intentional just like summary notifications for hold update actions not being there is intentional. Just checking.

correct, did not make any changes to holds intentionally, but checking with Ruth on that, asked her in the issue.

File: config/application.rb [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 require_relative 'boot'
 
-require 'rails'
+# require 'rails'
```
Comment (cdmo) [inline config/application.rb:5]: But... this doesn't do anything? It just removes a commented out line that shows the whole rails suite isn't included, or am I missing something?

Comment (banukutlu) [reply config/application.rb:5]: yeah yeah it does not, I was planning to go forward with rails upgrade to 6.1 and exclude activestorage but then realized rspec-rails not compatible with 6.1. I will get rid of this commit, then do a rebase if we get 6.0.3.6 in

**Review by cdmo — APPROVED** [review]
⛵

**Pull request merged**